docs: add securiy document (#333)
Kong/kubernetes-ingress-controller: docs/concepts/security.md

# Security

This document explains the security aspects of Kong Ingress Controller.

Kong Ingress Controller communicates with Kubernetes API-server and Kong's
Admin API. APIs on both sides offer authentication/authorization features
and the controller integrates with them gracefully.

## Kubernetes RBAC

Kong Ingress Controller is deployed with RBAC permissions as explained in
[deployment](deployment.md) document.
It has read and list permissions on most resources but requires update
and create permission for a few resources to provide seamless integration.
The permissions can be locked down further if needed depending on the specific
use-case.
This RBAC policy is associated with a ServiceAccount in Kubernetes, which

> takes of the Controller authenticating and authorizing against Kuberentes

Reviewer: Was this supposed to say "takes off"? I would rephrase it to avoid the gerunds. (Please clarify what it means first though.)

```suggestion
takes of the Controller authenticating and authorizing against Kuberentes
```

API-server.

> ## Kong Admin API protection

Reviewer: (If it's intentional, it's fine.)

```suggestion
## Kong Admin API Protection
```

Kong's Admin API is used to control configuration of Kong and proxying behavior.
If an attacker gets access to Kong's Admin API, Kong will be compromised

> and all bets are off at that point. Hence, it is important that the deployment

Reviewer: Instead of "all bets are off at that point," which does not sound sober/objective, describe what an attacker will be able to do. 

"An attacker will be able to perform the same actions as an authorized user--including actions that they may compromise the confidentiality, integrity, or availability of data proxied through Kong."

Something along those lines, more concise if possible.

> ensures that the likelyhood of this happening is minimized.

Reviewer: ```suggestion
ensures that the likelihood of this happening is minimized.
```

In the example deployements, the Controller and Kong's Admin API communicate

> over the loopback(lo) interface of the pod. There is no authorization or

Reviewer: ```suggestion
over the loopback (`lo`) interface of the pod. There is no authorization or
```

> authentication being done over the loopback listner.

Reviewer: ```suggestion
authentication taking place over the loopback listener.
```

Reviewer: if possible to say _what_ is not doing authentication or authorization, it'd be better to make the sentence active. 

"_X_ is not performing authentication or authorization over the loopback listener."

> Although not ideal, this setup is simple to get started and can be further

Reviewer: Describe in objective (and ideally measurable) terms how something is simple.

```suggestion
Although not ideal, this setup requires fewer steps to get started and can be further
```

hardened.

Please note that it is very important that Kong's Admin API is not accessible

> inside the cluster as any malicious service can change Kong's configuration.

Reviewer: just to check--did you mean outside instead of inside?

Author: Nope, inside is correct here.

What I mean to say:

Deploying Kong Admin API without any sort of authentication and open even inside a cluster or private network is a bad idea. You should always have authentication enabled on the Admin API or not expose it at all.

Reviewer: Understood, sounds good.

If you're exposing Kong's Admin API itself outside the cluster, please ensure
that you have the necessary authentication in place first.

### Authentication on Kong's Admin API

If Kong's Admin API is protected with one of the authentication plugins,
the Controller can authenticate itself against it to add another layer of
security.

> The controller comes with support for injecting arbitrary HTTP headers

Reviewer: ```suggestion
The Controller comes with support for injecting arbitrary HTTP headers
```

in the requests it makes to Kong's Admin API, which can be used to inject
authentication credentials.
The headers can be specified using the CLI flag `--admin-header` in the Ingress
Controller.

> Ingress Controller will support mutual TLS based authentication on Kong's Admin

Reviewer: Just for consistency, see lines above.

(Compound modifier needs hyphens.) 

```suggestion
The Ingress Controller will support mutual-TLS-based authentication on Kong's Admin
```

API in future.

### Kong Enterprise RBAC

Kong Enterprise comes with support for authentication and authorization on
Kong's Admin API.

> Once a RBAC token is provisioned, Kong Ingress Controller can use the RBAC

Reviewer: ```suggestion
Once an RBAC token is provisioned, Kong Ingress Controller can use the RBAC
```
